README.md
# Welcome to SPPMon and SPPCheck

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/5826/badge)](https://bestpractices.coreinfrastructure.org/projects/5826)

SPPMon is an open-source project initiated by the IBM Spectrum Protect Plus development team. The project's goal is to provide a monitoring system for IBM Spectrum Protect Plus that offers multiple options for a day to day monitoring of a data protection environment. The primary focus is the workflows of IBM Spectrum Protect Plus itself, such as job volume and duration or catalog protection, and the consumption of system resources like memory and CPU of all systems related to the data protection environment.
The built-in functionality to monitor the SPP server, vSnap, VADP, and Microsoft 365 proxy systems and can be expanded easily for other systems like application servers.

This introduction focuses on monitoring for IBM Spectrum Protect Plus. For information on monitoring IBM Spectrum Protect, see the documentation for [SPMon](https://github.com/IBM/sppmon/blob/master/spmon/README.md), which is currently in technical preview.

The SPPMon project consists of three major components.
The SPPMon core engine (open-source) is used to query the system data and ingest it into a database.
An Influx time-series database is used to store and prepare the collected data for the graphical interface.
Grafana is utilized as the graphical interface for the project.

SPPMon got extended in September 2022 by SPPCheck in the context of the master thesis of one of the primary developers.
SPPCheck is a system requirement verification and prediction tool aiming to enhance the existing functionality by verifying whether a system was set up correctly according to IBM's recommendations and predicting its future development.
It focuses on the storage consumption of all associated vSnaps and the server's memory and catalog space and is open to future expansion of its capabilities.\
SPPCheck re-uses the existing components and integrates SPPMons core engine while offering a PDF report besides the typical Grafana Dashboard.

The below picture describes the components and the general workflow on a high level.

![SPP / SPPmon Overview](https://github.com/IBM/sppmon/blob/master/pictures/sppmon_architecture.PNG)

## Structure of the Documentation

* The documentation is split into two sections: The user guide and the developer guide. The user guide includes all information needed to set up and configure the SPPMon system and data protection environment. The developer guide includes all information needed to improve the SPPMon system with more functionality and contribute to the open-source projects.

* SPPMon can be set up from [scratch on a Linux operating [system](https://github.com/IBM/spectrum-protect-sppmon/wiki/System-requirements) or deployed in a [containerized environment](https://github.com/IBM/spectrum-protect-sppmon/wiki/SPPmon-as-a-Container). See the [install options overview](https://github.com/IBM/spectrum-protect-sppmon/wiki/Install-overview) for more details and a distinction between each option.

* SPPCheck is automatically deployed when using the latest version of SPPMon and can be executed using its distinct [command line arguments](https://github.com/IBM/spectrum-protect-sppmon/wiki/SPPCheck-Command-line-Overview)

* Find the full documentation in the project [Wiki](https://github.com/IBM/spectrum-protect-sppmon/wiki)

### Short Links

* [Releases](https://github.com/IBM/spectrum-protect-sppmon/releases)
* [Changelog](https://github.com/IBM/spectrum-protect-sppmon/blob/master/CHANGELOG.md)
* [FAQ](https://github.com/IBM/spectrum-protect-sppmon/wiki/Frequently-asked-Questions)
* [Overview of required and created users](https://github.com/IBM/spectrum-protect-sppmon/wiki/Overview-of-users)
* [Bug Report and Feature Requests](https://github.com/IBM/spectrum-protect-sppmon/issues)
* [Installation of SPPMon](https://github.com/IBM/spectrum-protect-sppmon/wiki/Install-overview)
* [SPPMon Command line arguments](https://github.com/IBM/spectrum-protect-sppmon/wiki/SPPMon-Command-line-Overview)
* [SPPCheck Command line arguments](https://github.com/IBM/spectrum-protect-sppmon/wiki/SPPCheck-Command-line-Overview)
* [SPPCheck Export PDF Report](https://github.com/IBM/spectrum-protect-sppmon/wiki/SPPCheck-Export-PDF-Report)
* [General Wiki](https://github.com/IBM/spectrum-protect-sppmon/wiki)


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## SPPMon [Unreleased] - yyyy-mm-dd

### Added

### Changed

### Fixed

### Known Issues

## SPPCheck [Unreleased] - yyyy-mm-dd

### Added

### Changed

### Fixed

### Known Issues

## SPPMon [1.2.4] / SPPCheck [1.0.3] - 2023-02-23

### Added

* Added a `requirements.in` file for better versioning of dependencies

### Changed

* Added some macOS ignores to the gitignore
* Upgraded the dockerfile to a more secure standard
* Changed the requirements within the `requirements.txt` to be soft-locked, using `~=` to allow patch upgrades.

### Fixed

* Fixed security issues with the dependencies by updating them all to the latest state

## SPPMon [1.2.3] - 2022-12-05

### Changed

* Upgraded Dockerfile python version from 3.8 to 3.10

### Fixed

* Upgraded requirements file

### Known Issues

* There might be an issue with the alerts saved within the dashboard, requires investigation.

## SPPCheck [1.0.2] - 2022-12-05

### Added

* Debug information for accessing the excel json structure

### Fixed

* Removed another feature to allow comparability with python3.8 instead of 3.9
* Choosing the an automatic RP in report controller no longer debugs the keyword-content prefix, but the self defined variable rp_prefix
* Excel Json structure path within the excel controller has now the correct `sppCheck` case, as windows is case-insensitive
* Temporary directory path within the report controller has now the correct `sppCheck` case, as windows is case-insensitive
* Crash of Grafana-Panel downloader when using https fixed
* Upgraded requirements file

### Known Issues

## SPPMon [1.2.2] - 2022-08-30

### Changed

* Updated the requirements file to the latest versions

### Fixed

* Included pandas and numpy in the requirements file
* Removed Python 10 features, allowing compatibility with python 8.

## SPPCheck [1.0.1] - 2022-08-30

### Fixed

* Split up folder between `sppcheck` and `sppCheck` which was not shown on a windows development system
* Included pandas and numpy in the requirements file
* Removed Python 10 features, allowing compatibility with python 8.
* Fix of the default config file, removing an unwanted comma, causing both SPPMon and SPPCheck to fail.

## [1.2.1] - 2022-08-21

### Added

* Release of SPPCheck v1.0.0
  * Its changes will be logged within this file
  * SPPCheck function and internal structure
  * Added Grafana Dashboard for SPPCheck

### Changed

* Upgraded CodeQL to v3
* Extended `SppUtils.mk_logger_file` method by an logger_dir argument to allow different log files for SPPMon and SPPCheck
  * Added "sppcheckLogs" to the gitignore file.
* Moved all cSpell settings from the workspace file to the settings file
* Removed/Moved total count of exceptions during execution
* Moved verbose setting of sub-modules into init method to avoid issues and clarify program structure
* Error-Messages now have the prefix "ERROR: "
* Added InfluxClient function "get_list_rp" to query all retention policies, moving it out of existing functionality

### Fixed

* Code scanning alert #2 and #3: Printing passwords into logger due to an faulty if-expression.
* CheckPID file now uses the class-verbose setting and no longer ARGS-Verbose, making it independent of actual args.
* Fixed PID file checks under windows
* Fixed PID file entries not being deleted.
* Removed unnecessary and confusing check when creating a SelectionQuery

### Known Issues

* The Grafana Dashboards are missing the alerts. This is known and will be fixed within the next patch (see #115 )
* SPPCheck lacks a lot of documentation, which is to be delivered on a later date.

## [1.2.0] - 2022-06-17

### Added

* Adds FullLogs and LoadedSystem information to the Grafana SPPMon Runtime Duration panel.
* Adds clarification that all timestamps are changed into second precision on insert.
* Adds support for batch insert to insert into a different retention policy
* Selection queries now also support an alternative retention policy to query from
* CreateRP-Method in the influxClient to allow creating non-lasting retention policies
* Adds pandas-stubs, openpyxl and pyxlsb to requirements file
* Prints total count of errors during the execution if there are any - instead of "script finished" output
* Added type spelling dictionary to the settings file
* Added linting settings to the settings file
* Added two new tables to the definitions.py file which are only executed if SPPCheck is executed.

### Changed

* Moved pid files-functions and other functions from SPPMon to helper functions.
  * Adjusted log messages to make their message generic if required.
  * Added arguments to replace self-access.
* Changed default log and pid-file locations from `home/sppmonLogs/FILE` to `spectrum-protect-sppmon/sppmonLogs/FILE`
* Added init declarations in the SPPMon `__init__` function to allow overview of all self-vars
* Changed SelectionQueries to only take a single table instead of an List of such, as it isn't required and complicates the code
* Predefined retention policies are no longer protected in the definitions module.
* Sending a select query now raises a error when it fails.
* Allows SelectionQueries from table to be another inner-selection query
  * Changed all calls of the constructor
  * Made sure to only allow inner queries when using the selection-keyword.
* Changed the severity of unknown type-annotation due to submodules not being typed from error to warning.
* Minor printing edits when generifing Dashboard
* Removes deprecated functions

### Fixed

* Fixed typo in exception_utils.py file, renaming it and all references on it.
* Lots of typos in the inline documentation
* SelectionQuery and associated methods: Introduces Optional annotation to fix linter error

### Known Issues


python/mon.py
"""
 ----------------------------------------------------------------------------------------------
 (c) Copyright IBM Corporation 2022. All Rights Reserved.

 IBM Spectrum Protect Family Software

 Licensed materials provided under the terms of the IBM International Program
 License Agreement. See the Software licensing materials that came with the
 IBM Program for terms and conditions.

 U.S. Government Users Restricted Rights:  Use, duplication or disclosure
 restricted by GSA ADP Schedule Contract with IBM Corp.

 ----------------------------------------------------------------------------------------------
SPDX-License-Identifier: Apache-2.0

Description:
 The code in this branch ("SpMon") is part of an experimental project to
 refactor SppMon services for use in Spectrum Protect environments.

 It is based on the work of Niels Korchinsky, but it is not maintained or
 updated by him.

 This project is under development and is not currently considered ready
 for general use.

Repository:
  https://github.com/IBM/spectrum-protect-sppmon

Author (SppMon):
 Daniel Wendler
 Niels Korschinsky

Author (SpMon):
 Daniel Wendler
 Niels Korschinsky
 Daniel Boros
 James Damgar
 Rob Elder
 Sean Jones
 Raymond Shum

Changelog (SppMon):
 02/06/2020 version 0.2   tested with SPP 10.1.5
 02/13/2020 version 0.25  improved debug logging & url encoding
 02/20/2020 version 0.31  added function to add / modify REST API url parameters
                          added jobLogDetails function to capture joblogs and store in Influx DB
 03/24/2020 version 0.4   migrated to Influxdb
 04/15/2020 version 0.6   reworked all files with exception handling
 04/16/2020 version 0.6.1 Hotfixing vmStats table
 04/16/2020 version 0.6.2 Split into multiple tables
 04/17/2020 version 0.6.3 New Catalog Statistics and Reduced JobLogs store to only Summary.
 04/18/2020 version 0.6.4 Parsing two new JobLogID's
 04/20/2020 version 0.6.4.1 Minor change to jobLogs
 04/22/2020 version 0.6.5 Improved SSH Commands and added new Stats via ssh
 04/23/2020 version 0.7   New module structure
 04/27/2020 version 0.7.1 Fixes to index errors breaking the execution.
 04/27/2020 version 0.7.2 Reintroduced all joblogs and added --minimumLogs
 04/30/2020 version 0.8   Reworked Exception system, introduces arg grouping
 05/07/2020 version 0.8.1 Part of the documentation and typing system, renamed program args
 05/14/2020 version 0.8.2 Cleanup and full typing
 05/18/2020 version 0.9   Documentation finished and some bugfixes.
 05/19/2020 version 0.9.1 Moved future import into main file.
 06/02/2020 version 0.9.2 Fixed df ssh command, introduced CLOUDPROXY and shortened ssh.py file.
 06/03/2020 version 0.9.3 Introduces --hourly, grafana changes and small bugfixes
 07/16/2020 version 0.9.4 Shift of the --joblogs to --daily as expected
 07/16/2020 version 0.9.5 Dynamically shift of the pagesize for any kind of get-API requests
 08/02/2020 version 0.10.0 Introducing Retention Policies and Continuous Queries, breaking old tables
 08/25/2020 version 0.10.1 Fixes to Transfer Data, Parse Unit and Top-SSH-Command parsing
 09/01/2020 version 0.10.2 Parse_Unit fixes (JobLogs) and adjustments on timeout
 11/10/2020 version 0.10.3 Introduced --loadedSystem argument and moved --minimumLogs to deprecated
 12/07/2020 version 0.10.4 Included SPP 10.1.6 additional job information features and some bugfixes
 12/29/2020 version 0.10.5 Replaced ssh 'top' command by 'ps' command to bugfix truncating data
 01/22/2021 version 0.10.6 Removed `--processStats`, integrated in `--ssh` plus Server/vSnap `df` root recording
 01/22/2021 version 0.10.7 Replaced `transfer_data` by `copy_database` with improvements
 01/28/2021 version 0.11   Copy_database now also creates the database with RP's if missing.
 01/29/2021 version 0.12   Implemented --test function, also disabling regular setup on certain args
 02/09/2021 version 0.12.1 Hotfix job statistic and --test now also checks for all commands individually
 02/07/2021 version 0.13   Implemented additional Office365 Joblog parsing
 02/10/2021 version 0.13.1 Fixes to partial send(influx), including influxdb version into stats
 03/29/2021 version 0.13.2 Fixes to typing, reducing error messages and tracking code for NaN bug
 07/06/2021 version 0.13.3 Hotfixing version endpoint for SPP 10.1.8.1
 07/09/2021 version 0.13.4 Hotfixing storage exception, changing top-level exception handling to reduce the need of further hotfixes
 08/06/2021 version 0.13.5 Fixing PS having unintuitive CPU-recording, reintroducing TOP to collect CPU information only
 07/14/2021 version 0.13.6 Optimizing CQ's, reducing batch size and typo fix within cpuram table
 07/27/2021 version 0.13.7 Streamlining --test arg and checking for GrafanaReader on InfluxSetup
 08/02/2021 version 0.13.8 Enhancement and replacement of the ArgumentParser and clearer config-file error messages
 08/10/2021 version 0.13.9 Rework of the JobLogs and fix of Log-Filter.
 08/18/2021 version 0.14   Added install script and fixed typo in config file, breaking old config files.
 08/22/2021 version 0.15   Added --fullLogs argument and reduced regular/loaded joblog query to SUMMARY-Only
 08/25/2021 version 0.15.1 Replaced SLA-Endpoint by so-far unknown endpoint, bringing it in line with other api-requests.
 08/27/2021 version 1.0.0  Release of SPPMon
 08/27/2021 version 1.0.1  Reverted parts of the SLA-Endpoint change
 08/31/2021 version 1.0.2  Changed VADP table definition to prevent drop of false duplicates
 09/09/2021 version 1.1.0  Increase logging for REST-API errors, add ssh-client skip option for cfg file.
 02/22/2021 version 1.1.1  Only ssh-calls the vSnap-api if it is available
 06/17/2022 version 1.2.0  Change of logfile location, bug and documentation fixes. Removes deprecated functions.

"""
from __future__ import annotations

import functools
import sys

from argparse import ArgumentError, ArgumentParser
from spmon import SpMon

# Version:
SPP_VERSION = "1.2.0  (2022/06/17)"
SP_VERSION = "0.0.0 (2022/06/20)"
# ----------------------------------------------------------------------------
# command line parameter parsing
# ----------------------------------------------------------------------------

# Determine which mode this is running in: SP vs. SPP
mode = sys.argv[1]  # SP vs. SPP

# Parse the remaining arguments
parser = ArgumentParser(
    # exit_on_error=False, TODO: Enable in python version 3.9
    description=
    """Monitoring and long-term reporting for IBM Spectrum Protect and IBM Spectrum Protect Plus.
 Provides a data bridge from SP/SPP to InfluxDB and provides visualization dashboards via Grafana.

 This program provides functions to query IBM Spectrum Protect and IBM Spectrum Protect Plus Servers,
 VSNAP, VADP and other servers via REST API and ssh. This data is stored into a InfluxDB database.""",
    epilog="For feature-requests or bug-reports please visit https://github.com/IBM/spectrum-protect-sppmon")

# Applicable to both SP and SPP
parser.add_argument("--cfg", required=True, dest="configFile", help="REQUIRED: specify the JSON configuration file")
parser.add_argument("--verbose", dest="verbose", action="store_true", help="print to stdout")
parser.add_argument("--debug", dest="debug", action="store_true", help="save debug messages")
parser.add_argument("--test", dest="test", action="store_true", help="tests connection to all components")
parser.add_argument("--ssh", dest="ssh", action="store_true", help="execute monitoring commands via ssh")
parser.add_argument("--cpu", dest="cpu", action="store_true", help="capture SPP server CPU and RAM utilization")
parser.add_argument("--storages", dest="storages", action="store_true", help="store storages (vsnap) statistics")
parser.add_argument("--copy_database", dest="copy_database",
                    help="Copy all data from .cfg database into a new database, specified by `copy_database=newName`. Delete old database with caution.")
parser.add_argument("--constant", dest="constant", action="store_true",
                    help="execute recommended constant functions: (ssh, cpu, sppCatalog)")
parser.add_argument("--hourly", dest="hourly", action="store_true",
                    help="execute recommended hourly functions: (constant + jobs, vadps, storages)")
parser.add_argument("--daily", dest="daily", action="store_true",
                    help="execute recommended daily functions: (hourly +  joblogs, vms, slaStats, vmStats)")
parser.add_argument("--all", dest="all", action="store_true", help="execute all functions: (daily + sites)")

# SPP-specific arguments
if mode == "SPP":
    parser.add_argument("--jobs", dest="jobs", action="store_true", help="store job history")
    parser.add_argument("--jobLogs", dest="jobLogs", action="store_true",
                        help="retrieve detailed information per job (job-sessions)")
    parser.add_argument("--loadedSystem", dest="loadedSystem", action="store_true",
                        help="Special settings for loaded systems, increasing API-request timings.")
    parser.add_argument("--fullLogs", dest="fullLogs", action="store_true",
                        help="Requesting any kind of Joblogs instead of the default SUMMARY-Logs.")
    parser.add_argument("--vms", dest="vms", action="store_true", help="store vm statistics (hyperV, vmWare)")
    parser.add_argument("--vmStats", dest="vmStats", action="store_true",
                        help="calculate vm statistic from catalog data")
    parser.add_argument("--slaStats", dest="slaStats", action="store_true",
                        help="calculate vm's and applications per SLA")
    parser.add_argument("--vadps", dest="vadps", action="store_true", help="store VADPs statistics")
    parser.add_argument("--sites", dest="sites", action="store_true", help="store site settings")
    parser.add_argument("--sppcatalog", dest="sppcatalog", action="store_true",
                        help="capture Spp-Catalog Storage usage")

# SP-specific arguments
if mode == "SP":
    parser.add_argument("-v", '--version', action='version',
                        version="Spectrum Protect Monitoring (SPMon) version " + SP_VERSION)
    parser.add_argument("--queries", required=True, dest="queryFile", help="REQUIRED: SP summary record queries file")
else:
    parser.add_argument("-v", '--version', action='version',
                        version="Spectrum Protect Plus Monitoring (SPPMon) version " + SPP_VERSION)

print = functools.partial(print, flush=True)

# Define error codes
ERROR_CODE_START_ERROR: int = 3
ERROR_CODE_CMD_ARGS: int = 2
ERROR_CODE: int = 1
SUCCESS_CODE: int = 0

# Parse arguments
try:
    ARGS = parser.parse_args(sys.argv[2:])
except SystemExit as exit_code:
    if (exit_code.code != SUCCESS_CODE):
        print("> Error when reading arguments.", file=sys.stderr)
        print("> Please make sure to specify a config file and check the spelling of your arguments.", file=sys.stderr)
        print("> Use --help to display all argument options and requirements", file=sys.stderr)
    exit(exit_code)
except ArgumentError as error:
    print(error.message)
    print("> Error when reading arguments.", file=sys.stderr)
    print("> Please make sure to specify a config file and check the spelling of your arguments.", file=sys.stderr)
    print("> Use --help to display all argument options and requirements", file=sys.stderr)
    exit(ERROR_CODE_CMD_ARGS)

if __name__ == "__main__":
    # Run in one of the supported modes
    if mode == "SPP":
        # SppMon(ARGS).main()
        print("> SPP has been disconnected while developing SPMon", file=sys.stderr)
    elif mode == "SP":
        SpMon(ARGS).main()
